db/crud.go

```diff
@@ -1165,6 +1165,93 @@ func (db *DatabaseCollectionWithUser) PutExistingRevWithBody(ctx context.Context
 
 }
 
+func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Body, docID string) (*uint32, base.Set, channels.AccessMap, channels.AccessMap, error) {
+	doc := &Document{
+		ID:    docID,
+		_body: body,
+	}
+	oldDoc := doc
+	if docID != "" {
+		if docInBucket, err := db.GetDocument(ctx, docID, DocUnmarshalAll); err == nil {
+			oldDoc = docInBucket
+			if doc._body == nil {
+				body = oldDoc._body
+				doc._body = body
+			}
+			doc._body[BodyRev] = oldDoc.SyncData.CurrentRev
+		} else {
+			return nil, nil, nil, nil, err
+		}
+	}
+
+	delete(body, BodyId)
+
+	// Get the revision ID to match, and the new generation number:
+	matchRev, _ := body[BodyRev].(string)
+	generation, _ := ParseRevID(ctx, matchRev)
+	if generation < 0 {
+		return nil, nil, nil, nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid revision ID")
+	}
+	generation++
+	oldDoc.RevID = body.ExtractRev()
+
+	// Create newDoc which will be used to pass around Body
+	newDoc := &Document{
+		ID: docID,
+	}
+	// Pull out attachments
+	newDoc.DocAttachments = GetBodyAttachments(body)
+	delete(body, BodyAttachments)
+
+	delete(body, BodyRevisions)
+
+	err := validateAPIDocUpdate(body)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	bodyWithoutInternalProps, wasStripped := stripInternalProperties(body)
+	canonicalBytesForRevID, err := base.JSONMarshalCanonical(bodyWithoutInternalProps)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	// We needed to keep _deleted around in the body until we generated a rev ID, but now we can ditch it.
+	_, isDeleted := body[BodyDeleted]
+	if isDeleted {
+		delete(body, BodyDeleted)
+	}
+
+	// and now we can finally update the newDoc body to be without any special properties
+	newDoc.UpdateBody(body)
+
+	// If no special properties were stripped and document wasn't deleted, the canonical bytes represent the current
+	// body.  In this scenario, store canonical bytes as newDoc._rawBody
+	if !wasStripped && !isDeleted {
+		newDoc._rawBody = canonicalBytesForRevID
+	}
+
+	newRev := CreateRevIDWithBytes(generation, matchRev, canonicalBytesForRevID)
+	newDoc.RevID = newRev
+	mutableBody, metaMap, newRevID, err := db.prepareSyncFn(oldDoc, newDoc)
+	if err != nil {
+		base.InfofCtx(ctx, base.KeyDiagnostic, "Failed to prepare to run sync function: %v", err)
+		return nil, nil, nil, nil, err
+	}
+	if oldDoc.History != nil {
+		if err := oldDoc.History.addRevision(newDoc.ID, RevInfo{ID: newRev, Parent: matchRev, Deleted: isDeleted}); err != nil {
+			base.InfofCtx(ctx, base.KeyCRUD, "Failed to add revision ID: %s, for doc: %s, error: %v", newRev, base.UD(doc.ID), err)
+			return nil, nil, nil, nil, base.ErrRevTreeAddRevFailure
+		}
+	}
+
+	syncExpiry, _, channelSet, access, roles, err := db.runSyncFn(ctx, oldDoc, mutableBody, metaMap, newRevID, true)
```

Comment:
Looking at this again, it seems like a lot of the work you're having to do with the dryRun flag could be avoided if this function was calling the lower level DatabaseCollectionWithUser.ChannelMapper.MapToChannelsAndAccess directly (which actually does the sync function execution).  Do we need any of the other handling in runSyncFn or getChannelsAndAccess for the dry run?

I can see how checkDocChannelsAndGrantsLimits information would actually be useful for a dry run, but you've got that disabled at the moment anyway (since we don't want to warn/log).  Maybe call DatabaseCollectionWithUser.ChannelMapper.MapToChannelsAndAccess  and pass the results to a refactored checkDocChannelsAndGrantsLimits that will allow us to return the warnings in the SyncFnDryRun response?